### ISSUE TITLE
Editing distant LOD terrain?

### ISSUE THREAD

**Issue by 16millionoctopodes** [Zylann/godot_voxel#316]
Forgive me if I'm overlooking something obvious, but is there any way to edit distant LOD terrain, ie: voxels that are not within close range of a VoxelViewer?

**Comment by Zylann** [comment]
There is no way to edit chunks that are not loaded at the moment.
The very idea of how VoxelLodTerrain works is that it is a streamed system, this is what allows it to render massive distances. Just like you can't edit unloaded chunks in minecraft, you can't edit past LOD0 here. Anything beyond is mipped data which is only there to show an approximation instead of empty void, and editing approximated voxel data is a no-go.
The way to do it in the intented design is to place a `VoxelViewer` near the area you want to edit so it gets loaded and becomes editable. However this terrain does not support multi-viewers yet, so you have to move the first viewer there.

There is another design that could be explored in the future, where all the edited blocks are loaded at once. In such design, you would be able to edit anywhere, however the drawbacks would be:
- If the total amount of blocks does not fit in memory, it won't be possible to play it
- Due to the above, caching generated chunks will be more limited
- The world may have to be finite instead of the near-infinite boundaries that can be chosen currently

**Comment by 16millionoctopodes** [comment]
>The way to do it in the intented design is to place a VoxelViewer near the area you want to edit so it gets loaded and becomes editable. However this terrain does not support multi-viewers yet, so you have to move the first viewer there.

During the moving of the VoxelViewer, will the appearance of the terrain around the original location be affected? For example, let's say that normally the VoxelViewer is centered around the player camera. Some distance away, a meteor falls and makes a huge crater in the ground. So I move the VoxelViewer away from the player camera and place it at the impact site, make the crater edits, and then return the VoxelViewer back to the player camera. From the player's point of view, will the nearby terrain be "warped" to a lower resolution while the distant crater edits are being made?

**Comment by Zylann** [comment]
Yes, with a single viewer, the terrain will likely unload a bit and simplify around the player. This is one of the issues multi-viewer support could avoid.

**Comment by 16millionoctopodes** [comment]
Okay, I just want to be absolutely sure if I understand how LOD terrain works. Is this right?
- Full-resolution data for a Block of voxels _doesn't exist at all_ unless the viewer is within range, or otherwise specified by an external data source such as a region file, sqlite database entry, etc.
- The generator creates larger lower-resolution Blocks for distant terrain and smaller higher-resolution Blocks for nearby terrain.
- Whenever the viewer moves, approaching Blocks are octally subdivided and re-generated at smaller sizes and finer resolutions, while receding Blocks are octally combined and re-generated at larger sizes and rougher resolutions.
- Editing is only permitted on full-resolution Blocks, because editing a low-resolution Block would introduce unresolvable ambiguities as to how to correctly re-generate it for varying resolutions.

**Comment by Zylann** [comment]
> Full-resolution data for a Block of voxels doesn't exist at all unless the viewer is within range

Correct

> or otherwise specified by an external data source such as a region file, sqlite database entry, etc.

Not sure what you mean here, but in theory if access was immediately required regardless, there would be no other choice but to force a load of every block needed to access. Which in turn spawns the question of when to unload it etc.

> The generator creates larger lower-resolution Blocks for distant terrain and smaller higher-resolution Blocks for nearby terrain.

This is true, but it's not just about the generator. The generator only fills the gaps for blocks that were never user-made/altered. LOD works too with blocks that were modified, "mipmaps" are produced at every level of detail, such that if you dig a crater, that same crater remains visible from far away.

> Whenever the viewer moves, approaching Blocks are octally subdivided and re-generated at smaller sizes and finer resolutions, while receding Blocks are octally combined and re-generated at larger sizes and rougher resolutions.

Correct, although usually receding blocks don't need to re-generate because their "parent" level is still in memory. Loading only occurs if the "child" is no longer in memory, which is more likely.

> Editing is only permitted on full-resolution Blocks, because editing a low-resolution Block would introduce unresolvable ambiguities as to how to correctly re-generate it for varying resolutions.

Yes.

**Comment by 16millionoctopodes** [comment]
> Not sure what you mean here

Just a crazy idea I was toying with. Region files seem to contain all of a Block's LODs, so I was thinking that maybe I could redirect any commands for editing distant blocks to separate functions that would create region files / database entries for the targeted Blocks (if the relevant files/entries didn't already exist), make the edits to the relevant files/entries directly, and then force the game to re-create the associated Blocks based on the edited files/entries. But I'm probably not going to actually do this, since it's clumsy, messy, and I clearly still don't completely understand how the module works.

**Comment by Zylann** [comment]
I've been willing to give a try to the approach I mentionned in https://github.com/Zylann/godot_voxel/issues/316#issuecomment-929267950 for a while though, despite the downsides:

https://user-images.githubusercontent.com/1311555/135678650-90b09b2b-77ef-40d7-a46c-aa0f0c58dc38.mp4
